lib/northbound_grpc.cpp

```diff
@@ -37,1315 +39,1276 @@
+template <typename Q, typename S> class NewRpcState : RpcStateBase
 {
+	typedef void (frr::Northbound::AsyncService::*reqfunc_t)(
+		::grpc::ServerContext *, Q *,
+		::grpc::ServerAsyncResponseWriter<S> *,
+		::grpc::CompletionQueue *, ::grpc::ServerCompletionQueue *,
+		void *);
+	typedef void (frr::Northbound::AsyncService::*reqsfunc_t)(
+		::grpc::ServerContext *, Q *, ::grpc::ServerAsyncWriter<S> *,
+		::grpc::CompletionQueue *, ::grpc::ServerCompletionQueue *,
+		void *);
+
       public:
-	RpcState(NorthboundImpl *svc,
-		 void (NorthboundImpl::*cb)(RpcState<Q, S> *))
-	    : callback(cb), responder(&ctx), async_responder(&ctx),
-	      service(svc){};
+	NewRpcState(Candidates *cdb, reqfunc_t rfunc,
+		    void (*cb)(NewRpcState<Q, S> *), const char *name)
+	    : requestf(rfunc), callback(cb), responder(&ctx),
+	      async_responder(&ctx), name(name), cdb(cdb){};
+	NewRpcState(Candidates *cdb, reqsfunc_t rfunc,
+		    void (*cb)(NewRpcState<Q, S> *), const char *name)
+	    : requestsf(rfunc), callback(cb), responder(&ctx),
+	      async_responder(&ctx), name(name), cdb(cdb){};
+
+	CallState doCallback() override
+	{
+		CallState enter_state = this->state;
+		CallState new_state;
+		if (enter_state == FINISH) {
+			grpc_debug("%s RPC FINISH -> DELETED", name);
+			new_state = FINISH;
+		} else {
+			grpc_debug("%s RPC: %s -> PROCESS", name,
+				   call_states[this->state]);
+			new_state = PROCESS;
+		}
+		/*
+		 * We are either in state CREATE, MORE or FINISH. If CREATE or
+		 * MORE move back to PROCESS, otherwise we are cleaning up
+		 * (FINISH) so leave it in that state. Run the callback on the
+		 * main threadmaster/pthread; and wait for expected transition
+		 * from main thread. If transition is to DELETED delete us.
+		 */
+		this->state = new_state;
+		thread_add_event(main_master, c_callback, (void *)this, 0,
+				 NULL);
+		pthread_mutex_lock(&this->cmux);
+		while (this->state == new_state)
+			pthread_cond_wait(&this->cond, &this->cmux);
+		pthread_mutex_unlock(&this->cmux);
+
+		if (this->state == DELETED) {
+			grpc_debug("%s RPC: -> [DELETED]", name);
+			delete this;
+			return DELETED;
+		}
+		return this->state;
+	}
 
-	void doCallback() override
+	void do_request(::frr::Northbound::AsyncService *service,
+			::grpc::ServerCompletionQueue *cq) override
 	{
-		(service->*callback)(this);
+		grpc_debug("%s, posting a request for: %s", __func__, name);
+		if (requestf) {
+			NewRpcState<Q, S> *copy =
+				new NewRpcState(cdb, requestf, callback, name);
+			(service->*requestf)(&copy->ctx, &copy->request,
+					     &copy->responder, cq, cq, copy);
+		} else {
+			NewRpcState<Q, S> *copy =
+				new NewRpcState(cdb, requestsf, callback, name);
+			(service->*requestsf)(&copy->ctx, &copy->request,
+					      &copy->async_responder, cq, cq,
+					      copy);
+		}
+	}
+
+
+	static int c_callback(struct thread *thread)
+	{
+		auto _tag = static_cast<NewRpcState<Q, S> *>(thread->arg);
+		pthread_mutex_lock(&_tag->cmux);
```

Comment:
Could we have a comment here about the lock management here? I think you're right that you need to hold the lock, but it would be easier to follow if it were more explicit.

Reply:
I've updated the comment with description of the lock.